feat(subagent): add Phase 1 async enhancements (subprocess mode, hook notifications, batch)

TimeToBuildBob opened the pull request:
## Summary

Implements Phase 1 of the async subagents design from Issue #554.

## Changes

### 1. Subprocess Execution Mode
- New `use_subprocess=True` parameter for output isolation
- Runs gptme as subprocess, captures stdout/stderr separately

### 2. Callback Hooks
- `on_complete`: Called when subagent finishes with result
- `on_progress`: Called for progress updates (infrastructure for Phase 2)

### 3. Batch Execution Helper
- New `subagent_batch()` for fire-and-gather pattern
- `BatchJob` class with `wait_all()`, `is_complete()`, `get_completed()`

### 4. Enhanced Subagent Dataclass
- `execution_mode` field: 'thread' or 'subprocess'
- Callback fields for async notifications

### 5. Updated `subagent_wait()`
- Handles both thread and subprocess modes
- Configurable `timeout` parameter

## Design Document
See: [async-subagents-design.md](https://github.com/ErikBjare/bob/blob/master/knowledge/technical-designs/async-subagents-design.md)

## Tests
- All 15 existing tests pass
- Added 5 new tests for Phase 1 features

Ref: #554
<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Enhances subagent functionality with subprocess execution, batch processing, and hook-based notifications, adding new classes and tests for robust async operations.
> 
>   - **Behavior**:
>     - Adds `use_subprocess=True` parameter in `subagent()` for subprocess execution mode, isolating output.
>     - Introduces `subagent_batch()` for batch execution, managing multiple subagents concurrently.
>     - Implements `subagent_completion` hook for completion notifications via `LOOP_CONTINUE`.
>   - **Classes and Functions**:
>     - Adds `BatchJob` class with methods `wait_all()`, `is_complete()`, and `get_completed()` in `subagent.py`.
>     - Updates `Subagent` class to support both thread and subprocess modes.
>     - Modifies `subagent_wait()` to handle subprocess mode with a configurable `timeout`.
>   - **Tests**:
>     - Adds tests for subprocess mode, batch execution, and hook notifications in `test_tools_subagent.py`.
>     - Ensures subprocess creation and command construction are correct.
>     - Verifies hook-based completion notifications and batch job functionality.
>   - **Misc**:
>     - Registers `subagent_completion` hook in `subagent.py` for asynchronous notifications.
>     - Updates `conf.py` to include new subagent types in nitpick ignore list.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=gptme%2Fgptme&utm_source=github&utm_medium=referral)<sup> for 351c217a2ccf3fe24fabad2594055b1cf95fc530. You can [customize](https://app.ellipsis.dev/gptme/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

ErikBjare commented:
Looks interesting but not sure it quite succeeds in what we were trying to achieve, see my comment here: https://github.com/gptme/gptme/issues/554#issuecomment-3665579912

ErikBjare commented on gptme/tools/subagent.py:
```diff
@@ -41,23 +49,49 @@ class ReturnType:
 
 
-@dataclass(frozen=True)
+@dataclass
```
I don't think you should remove the frozen=True. Find some other way to communicate (attach a mutable ref or something) instead of mutating the object.

TimeToBuildBob commented:
🔄 **CI Rerun Triggered**

The failing `Test with openai/gpt-4o-mini` job failed due to a GitHub Actions infrastructure issue:

```
System.IO.IOException: No space left on device
```

The test suite itself **passed successfully** (✓ Run tests), but the runner ran out of disk space during cleanup. This is a transient infrastructure issue, not a code problem.

I've triggered a rerun of the failed job which should pass on a fresh runner with available disk space.

ErikBjare commented on gptme/tools/subagent.py:
```diff
@@ -350,26 +482,84 @@ def random_string(n):
+            logdir=logdir,
+            model=model_name,
+            workspace=workspace,
+        )
```
Valid concern, should fix. Or at least make follow-up issue about streamlining context definition for the core CLI and subagents (tools, system prompt (parts, special additions like return/update/ask-parent-agent instructions), output schema as mentioned in https://github.com/gptme/gptme/pull/962/changes#r2627245143, etc)

TimeToBuildBob commented:
## Thread Safety Fix ✅

Addressed the review comment about synchronizing access to `_subagent_results`:

**Commit** a65c407b3:
- Added `_subagent_results_lock = threading.Lock()` at module level
- Protected read access in `Subagent.status()` with lock
- Protected write access in `_monitor_subprocess()` with lock

This ensures thread-safe access when multiple subagents run in parallel and access the shared result cache.

All 20 subagent tests pass ✅

ErikBjare reviewed (changes requested):
Looks pretty good now, just some concerns about poor testing. Once you address those I will do some manual user-testing.

I see subagents cannot send "update" notifications or ask the top-level agent for clarifying info. This might be phase 2/3 work, but I think we should document it in docstrings and such already to make the design intent clear.

TimeToBuildBob commented:
Addressed the review feedback! ✅

**Documentation for Phase 2/3 design intent:**
- Added module docstring documenting Phase 1 implementation and future design intent
- Explains that subagents currently cannot send progress updates or request clarification
- Documents bidirectional communication plans for Phase 2/3

**Improved test coverage (4 new tests):**
- `test_subagent_status_returns_dict`: Tests subagent_status() returns valid dict
- `test_subagent_status_unknown_agent`: Tests ValueError for unknown agent
- `test_subagent_wait_basic`: Tests subagent_wait() can wait for completion
- `test_subagent_read_log_returns_string`: Tests subagent_read_log() returns string

Total tests: 20 → 24 (all passing)

Commit: f5ce4247f

ErikBjare commented on gptme/hooks/subagent_completion.py:
This file is now unused? Should be removed.

TimeToBuildBob commented:
## Review Comments Addressed ✅

@ErikBjare - Addressed your review comments in commit cb7fe88:

### 1. Removed unused hook file
✅ Deleted `gptme/hooks/subagent_completion.py` - hook functionality is now in `gptme/tools/subagent.py` via ToolSpec (as documented in my earlier commit 1748d53).

### 2. Added actual subprocess execution tests
✅ Added 5 new integration tests that actually execute subprocess mode (not mocked):
- `test_subprocess_actual_process_creation`: Verifies real subprocess.Popen object is created
- `test_subprocess_command_includes_required_flags`: Verifies gptme CLI flags (-n, --no-confirm, --logdir, --model)
- `test_subprocess_working_directory`: Verifies workspace cwd is set correctly
- `test_subprocess_full_flow_with_subagent_function`: Full flow test using subagent() function
- `test_subprocess_monitor_thread_started`: Verifies monitor thread starts for subprocess

**Test Results**: All 32 tests pass (was 27)

---

Ready for re-review!

TimeToBuildBob commented on gptme/tools/subagent.py:
```diff
@@ -2,13 +2,30 @@
+- **Bidirectional communication**: Establish message channels between parent/child for collaborative problem-solving
+- **Hierarchical coordination**: Support multi-level agent hierarchies with message routing and result aggregation
+
+These features require hook system integration and coordination protocol design.
```
✅ Addressed in 1748d5337

Moved the completion hook into the tool via ToolSpec pattern. The docstring now reflects the integrated hook behavior.

ErikBjare commented:
@greptileai review

TimeToBuildBob commented:
## ✅ Greptile Review Feedback Addressed

Commit 351c217a2 addresses all 3 Greptile review comments:

### 1. Memory optimization in subprocess monitoring (line 417)
- Changed from `.communicate()` to `.wait()` to avoid reading all output into memory
- Now relies on log file for result extraction (which is the canonical source for subprocess mode)
- Prevents memory issues with long-running subagents

### 2. Subprocess cleanup in test fixture (line 141)
- Added subprocess termination logic to `cleanup_subagents_after` fixture
- Graceful termination with 5s timeout, falls back to `.kill()` if needed
- Ensures clean test isolation for subprocess mode tests

### 3. Exception handling in thread mode (line 646)
- Wrapped `_create_subagent_thread()` in try-except block
- If execution fails, properly notifies with 'error' status instead of running notification with incomplete state
- Prevents misleading status reports from failed subagents

All 34 subagent tests pass ✅

Note: Bot review threads show as 'unresolved' due to GitHub's review system - only maintainers can manually resolve bot reviews.
